File: src/lib/trusted-device.ts
import { db } from "@/lib/drizzle";
import { trustedDevices } from "@/db/schema";
import { eq, and, gt } from "drizzle-orm";
import crypto from "crypto";
import { env } from "./env";

function getSecretKey(): string {
  return env.trustedDeviceSecret;
}

function hashToken(rawToken: string): string {
  return crypto.createHmac("sha256", getSecretKey()).update(rawToken).digest("hex");
}

function parseUserAgent(userAgent?: string): string {
  if (!userAgent) return "Unknown Device";
  let os = "Unknown OS";
  if (userAgent.includes("Windows")) os = "Windows";
  else if (userAgent.includes("Mac OS")) os = "macOS";
  else if (userAgent.includes("Linux")) os = "Linux";
  else if (userAgent.includes("Android")) os = "Android";
  else if (userAgent.includes("iPhone") || userAgent.includes("iPad")) os = "iOS";

  let browser = "Browser";
  if (userAgent.includes("Edg/")) browser = "Edge";
  else if (userAgent.includes("Chrome/")) browser = "Chrome";
  else if (userAgent.includes("Firefox/")) browser = "Firefox";
  else if (userAgent.includes("Safari/") && !userAgent.includes("Chrome/")) browser = "Safari";

  return `${browser} on ${os}`;
}

export async function createTrustedDevice(
  userId: string,
  userAgent?: string,
  ipAddress?: string
): Promise<string> {
  const rawToken = crypto.randomBytes(32).toString("hex");
  const tokenHash = hashToken(rawToken);
  const deviceName = parseUserAgent(userAgent);
  
  const expiresAt = new Date();
  expiresAt.setDate(expiresAt.getDate() + 30); // 30 days from now

  const [inserted] = await db
    .insert(trustedDevices)
    .values({
      userId,
      tokenHash,
      deviceName,
      ipAddress: ipAddress ?? null,
      lastUsedAt: new Date(),
      expiresAt,
    })
    .returning({ id: trustedDevices.id });

  // Cookie payload format: {id}.{rawToken}
  return `${inserted.id}.${rawToken}`;
}

export async function verifyTrustedDevice(
  userId: string,
  cookieValue: string | undefined
): Promise<boolean> {
  if (!cookieValue || !userId) return false;

  const parts = cookieValue.split(".");
  if (parts.length !== 2) return false;

  const [deviceId, rawToken] = parts;
  if (!deviceId || !rawToken) return false;

  const tokenHash = hashToken(rawToken);
  const now = new Date();

  try {
    const records = await db
      .select()
      .from(trustedDevices)
      .where(
        and(
          eq(trustedDevices.id, deviceId),
          eq(trustedDevices.userId, userId),
          eq(trustedDevices.tokenHash, tokenHash),
          gt(trustedDevices.expiresAt, now)
        )
      )
      .limit(1);

    if (records.length === 0) {
      return false;
    }

    // Update lastUsedAt asynchronously
    db.update(trustedDevices)
      .set({ lastUsedAt: now })
      .where(eq(trustedDevices.id, deviceId))
      .catch((err) => console.error("[TRUSTED DEVICE] Failed to update lastUsedAt:", err));

    return true;
  } catch (err) {
    console.error("[TRUSTED DEVICE] Verification error:", err);
    return false;
  }
}

export async function revokeTrustedDevice(deviceId: string, userId: string): Promise<boolean> {
  try {
    await db
      .delete(trustedDevices)
      .where(and(eq(trustedDevices.id, deviceId), eq(trustedDevices.userId, userId)));
    return true;
  } catch (err) {
    console.error("[TRUSTED DEVICE] Revoke error:", err);
    return false;
  }
}

export async function revokeAllUserDevices(userId: string): Promise<boolean> {
  try {
    await db.delete(trustedDevices).where(eq(trustedDevices.userId, userId));
    return true;
  } catch (err) {
    console.error("[TRUSTED DEVICE] Revoke all error:", err);
    return false;
  }
}

File: backend/internal/auth/trusted_device.go
package auth

import (
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"time"

	"github.com/google/uuid"

	"pusdatin/backend/internal/database"
)

type TrustedDeviceService struct {
	Store *database.Store
	Secret []byte
}

func NewTrustedDeviceService(store *database.Store, secret string) *TrustedDeviceService {
	return &TrustedDeviceService{Store: store, Secret: []byte(secret)}
}

func (s *TrustedDeviceService) hashToken(raw string) string {
	mac := hmac.New(sha256.New, s.Secret)
	mac.Write([]byte(raw))
	return hex.EncodeToString(mac.Sum(nil))
}

// parseUserAgent mirrors lib/trusted-device.ts parseUserAgent.
func parseUserAgent(ua string) string {
	if ua == "" {
		return "Unknown Device"
	}
	os := "Unknown OS"
	switch {
	case strings.Contains(ua, "Windows"):
		os = "Windows"
	case strings.Contains(ua, "Mac OS"):
		os = "macOS"
	case strings.Contains(ua, "Linux"):
		os = "Linux"
	case strings.Contains(ua, "Android"):
		os = "Android"
	case strings.Contains(ua, "iPhone"), strings.Contains(ua, "iPad"):
		os = "iOS"
	}
	browser := "Browser"
	switch {
	case strings.Contains(ua, "Edg/"):
		browser = "Edge"
	case strings.Contains(ua, "Chrome/"):
		browser = "Chrome"
	case strings.Contains(ua, "Firefox/"):
		browser = "Firefox"
	case strings.Contains(ua, "Safari/") && !strings.Contains(ua, "Chrome/"):
		browser = "Safari"
	}
	return browser + " on " + os
}

// Create returns the cookie value `{id}.{rawToken}` and persists the hash.
func (s *TrustedDeviceService) Create(ctx context.Context, userID, userAgent, ipAddress string) (string, error) {
	raw := make([]byte, 32)
	if _, err := rand.Read(raw); err != nil {
		return "", err
	}
	rawToken := hex.EncodeToString(raw)
	id := uuid.NewString()
	deviceName := parseUserAgent(userAgent)

	var ipPtr any
	if ipAddress != "" {
		ipPtr = ipAddress
	}
	if err := s.Store.CreateTrustedDeviceWithIP(ctx, id, userID, deviceName, s.hashToken(rawToken), time.Now().AddDate(0, 0, 30), ipPtr); err != nil {
		return "", err
	}
	return id + "." + rawToken, nil
}

// Verify checks the cookie token against the DB and refreshes last_used_at.
func (s *TrustedDeviceService) Verify(ctx context.Context, userID, cookieValue string) bool {
	if cookieValue == "" || userID == "" {
		return false
	}
	parts := strings.Split(cookieValue, ".")
	if len(parts) != 2 {
		return false
	}
	id, rawToken := parts[0], parts[1]
	if id == "" || rawToken == "" {
		return false
	}
	dev, err := s.Store.GetValidTrustedDevice(ctx, id, userID)
	if err != nil {
		return false
	}
	if !hmac.Equal([]byte(dev.TokenHash), []byte(s.hashToken(rawToken))) {
		return false
	}
	_ = s.Store.TouchTrustedDevice(ctx, id)
	return true
}

func (s *TrustedDeviceService) Revoke(ctx context.Context, deviceID, userID string) error {
	return s.Store.RevokeTrustedDevice(ctx, deviceID, userID)
}

func (s *TrustedDeviceService) RevokeAll(ctx context.Context, userID string) error {
	return s.Store.RevokeAllTrustedDevices(ctx, userID)
}
